src/sctools/bam.py

```diff
@@ -0,0 +1,125 @@
+        valid_chromosomes = [str(i) for i in range(1, 23)] + ['M', 'X', 'Y']
+        if isinstance(chromosome, int) and chromosome < 23:
+            chromosome = str(chromosome)  # try to convert
+        if chromosome not in valid_chromosomes:
+            raise ValueError('chromsome %s not valid. Must be one of %r' %
+                             (chromosome, valid_chromosomes))
+
+        with pysam.AlignmentFile(self._file, self._open_mode) as fin:
+            specific, nonspecific = 0, 0  # counters
+            chromosome = str(chromosome)
+            chromosome_indices = []
+            other_indices = []
+
+            for i, record in enumerate(fin):
+
+                if not record.is_unmapped:  # record is mapped
+                    if chromosome in record.reference_name and specific < n_specific:
+                        chromosome_indices.append(i)
+                        specific += 1
+                    elif nonspecific < include_other:
+                        other_indices.append(i)
+                        nonspecific += 1
+                elif nonspecific < include_other:  # record is not mapped
+                    other_indices.append(i)
+                    nonspecific += 1
+
+                # check termination condition (we have the requisite number of reads
+                if specific == n_specific and nonspecific == include_other:
+                    break
+
+        if specific < n_specific or nonspecific < include_other:
+            print('Warning: Only %d unaligned and %d reads aligned to chromosome %s were found in' 
+                  '%s' % (nonspecific, specific, chromosome, self._file))
+
+        if include_other != 0:
+            return chromosome_indices, other_indices
+        else:
+            return chromosome_indices
+
+
+class Tagger:
+    """Add tags to a bam file from tag generators
+
+    :method tag: add tags from an arbitrary number of tag generators. Typical use is to extract
+      barcodes from fastq files.
+    """
+
+    def __init__(self, bam_file):
+        """
+        :param str bam_file: location of bam file
+        """
+        if not isinstance(bam_file, str):
+            raise TypeError('bam_file must be type str, not %s' % type(bam_file))
+        self.bam_file = bam_file
+
+    def tag(self, output_bam_name, tag_generators):
+        """
+        given a bam file and tag generators in the same order, adds tags to the .bam file,
+        writes the resulting file to output_bam_name.
+
+        :param str output_bam_name: name of output tagged bam.
+        :param [fastq.TagGenerator] tag_generators: generators that yield Tag objects
+          (see fastq.Tag)
+        """
+        inbam = pysam.AlignmentFile(self.bam_file, 'rb', check_sq=False)
+        outbam = pysam.AlignmentFile(output_bam_name, 'wb', header=inbam.header)
+        try:
+            # zip up all the iterators
+            for *tag_sets, sam_record in zip(*tag_generators, inbam):
```

Comment:
I'm not familiar with pysam. I'm assuming this does not pull all reads in the bam into memory at once, although naively that's that it looks like. Can you confirm?

Reply:
Confirmed; it's an iterator, and pysam is the python wrapper for htslib. 

This line is zipping up three iterators that are lazily read. You get 4 fastq lines and one sam line per iteration.